src/main/java/net/minestom/server/inventory/AbstractInventory.java

```diff
@@ -0,0 +1,366 @@
+     */
+    public List<ItemStack> takeItemStacks(@NotNull List<ItemStack> itemStacks) {
+        List<ItemStack> result = new ArrayList<>();
+        itemStacks.forEach(itemStack -> {
+            if (!takeItemStack(itemStack)) {
+                result.add(itemStack);
+            }
+        });
+        return result;
+    }
+
+    /**
+     * Checks whether {@link ItemStack} can be fully taken from the inventory.
+     *
+     * @param itemStack the item to be checked
+     * @return true if the item can be fully taken from the inventory, false otherwise
+     */
+    public boolean canTakeItemStack(@NotNull ItemStack itemStack) {
+        final StackingRule stackingRule = itemStack.getStackingRule();
+        int amountLeft = stackingRule.getAmount(itemStack);
+        for (int i = 0; i < getInnerSize(); i++) {
+            ItemStack inventoryItem = getItemStack(i);
+            if (inventoryItem.isAir()) {
+                continue;
+            }
+            if (stackingRule.canBeStacked(itemStack, inventoryItem)) {
+                final int itemAmount = stackingRule.getAmount(inventoryItem);
+                if (amountLeft <= itemAmount) {
+                    return true;
+                }
+                amountLeft -= itemAmount;
+            }
+        }
+        return false;
+    }
+
+    /**
+     * Checks whether {@link ItemStack}s can be fully taken from the inventory.
+     *
+     * @param itemStacks items to be checked
+     * @return true if all the items can be fully taken from the inventory, false otherwise
+     */
+    public boolean canTakeItemStacks(@NotNull List<ItemStack> itemStacks) {
+        return itemStacks.stream().allMatch(this::canTakeItemStack);
+    }
+
+    public synchronized void replaceItemStack(int slot, @NotNull UnaryOperator<@NotNull ItemStack> operator) {
+        var currentItem = getItemStack(slot);
+        setItemStack(slot, operator.apply(currentItem));
+    }
+
+    /**
+     * Clears the inventory and send relevant update to the viewer(s).
+     */
+    public synchronized void clear() {
+        // Clear the item array
+        Arrays.fill(itemStacks, ItemStack.AIR);
+        // Send the cleared inventory to viewers
+        update();
+    }
+
+    public abstract void update();
+
+    /**
+     * Gets the {@link ItemStack} at the specified slot.
+     *
+     * @param slot the slot to check
+     * @return the item in the slot {@code slot}
+     */
+    @NotNull
+    public ItemStack getItemStack(int slot) {
+        return itemStacks[slot];
+    }
+
+    /**
+     * Gets all the {@link ItemStack} in the inventory.
+     * <p>
+     * Be aware that the returned array does not need to be the original one,
+     * meaning that modifying it directly may not work.
+     *
+     * @return an array containing all the inventory's items
+     */
+    @NotNull
+    public ItemStack[] getItemStacks() {
+        return itemStacks.clone();
```

Comment:
If we store the `ItemStack`s as a list, we could just return an unmodifiable view of the stack. Saves a clone.

Reply:
It does currently use an array as the size is final, I do not think that cloning is a huge concern

Reply:
Well the size could be set on a list, but this isn't a massive issue ig